src/translation-assistant.js

```diff
@@ -13,6 +13,10 @@ const MATH_REGEX = /\$(\\\$|[^\$])+\$/g;
 // e.g. ![](web+graphie://ka-perseus-graphie.s3.amazonaws.com/542f2b4e297910eed545a5c29c3866918655bab4)
 const GRAPHIE_REGEX = /\!\[\]\([^)]+\)/g;
 
+// Matches image link strings,
+// e.g. 'https://ka-perseus-graphie.s3.amazonaws.com/e75c49cb5753492629016169933ab63af3b9f122.png
+const IMAGE_REGEX = /https:\/\/[^\s]+\.png/g;
```

Comment:
I'm wondering if it would also make sense to handle `web+graphie://...` image links as well?

In this exercise for example: https://www.khanacademy.org/translations/edit/pl/science/physics/one-dimensional-motion/displacement-velocity-time/e/average-velocity-and-average-speed-exercise/by-pattern

There's a mix of `http://...png` and `web+graphie://...` image links, linked using the Image Widget in the exercise editor:

![screen shot 2018-01-30 at 9 46 27 am](https://user-images.githubusercontent.com/4722598/35572462-aa3ce3f8-05a2-11e8-8291-5140bcdfe7bb.png)

Changing `IMAGE_REGEX` to:

```
// Matches image link strings,
// e.g. https://ka-perseus-graphie.s3.amazonaws.com/e75c49cb5753492629016169933ab63af3b9f122.png,
// or web+graphie://ka-perseus-graphie.s3.amazonaws.com/542f2b4e297910eed545a5c29c3866918655bab4
// (if not a graphie string, ie. ![](web+graphie://...)).
const IMAGE_REGEX = /https:\/\/[^\s]+\.png|web\+graphie:\/\/[^\s\)]+/g;
```

Would group both types of links as one pattern:

![screen shot 2018-01-30 at 9 37 47 am](https://user-images.githubusercontent.com/4722598/35572669-1e906fae-05a3-11e8-8ba1-65f1e1efd63d.png)

What do you think?

Reply:
A good find for those `web+graphie://...` links! I wasn't aware of that use. Makes sense to extend current image pattern to include pure Graphie Links as well.

I had one problem while trying to make it work. This link is actually an exact subset of a Graphie Widget string, so regular expression would find Graphie Link within Graphie Widget as well. I tried using [negative lookbehind](https://www.regular-expressions.info/lookaround.html#lookbehind) to avoid a match in case of that `![](` prefix, but this regex feature is [not supported by JavaScript](https://www.regular-expressions.info/refadv.html). I ended up using positive lookahead:

- `https:[^\s]+\.png|web\+graphie:[a-z0-9\.\-/]+(?=[\s,]|$)`

Complication is consequence of these changes also supporting strings that contain other text in addition to links. Tests added cover these cases.

Reply:
Awesome, thank you for updating the regex.

> I had one problem while trying to make it work, though. This Graphie Link is actually an exact subset of a Graphie Widget string, so regular expression would find Graphie Link within Graphie Widget as well

Indeed, I think this issue stems from how `getMapping` works.

It's running `Math`, `Graphie`, `Image` and `Widget` regex over the input & output strings in turn, resulting in the `Image` regex throwing an error if it's not handling a string with graphie link inside `![]()`.

I think this is the only function in the library that expects the regex to handle strings in their "original" form.

`createTemplate` & `populateTemplate` apply regex in turn to mutate the string so by the time the `Image` regex is applied, the `Graphie` regex transformed the string to replace all `![](web+graphie://...)` to be `__GRAPHIE__`.